pkg/client/client_v2.go

```diff
@@ -0,0 +1,543 @@
+	// Meta is any extra metadata information about the version.
+	Meta map[string]string
+}
+
+// V2 manages the state and communication to the Elastic Agent over the V2 control protocol.
+type V2 interface {
+	// Start starts the connection to Elastic Agent.
+	Start(ctx context.Context) error
+	// Stop stops the connection to Elastic Agent.
+	Stop()
+	// UnitChanges returns channel client send unit change notifications to.
+	//
+	// User of this client must read from this channel, or it will block the client.
+	UnitChanges() <-chan UnitChanged
+	// Errors returns channel of errors that occurred during communication.
+	//
+	// User of this client must read from this channel, or it will block the client.
+	Errors() <-chan error
+	// Artifacts returns the artifacts client.
+	Artifacts() ArtifactsClient
+}
+
+// clientV2 manages the state and communication to the Elastic Agent over the V2 control protocol.
+type clientV2 struct {
+	target string
+	opts   []grpc.DialOption
+	token  string
+
+	versionInfo     VersionInfo
+	versionInfoSent bool
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+	client proto.ElasticAgentClient
+	cfgMu  sync.RWMutex
+	obsMu  sync.RWMutex
+
+	kickCh  chan struct{}
+	errCh   chan error
+	unitsCh chan UnitChanged
+	unitsMu sync.RWMutex
+	units   []*Unit
+
+	storeClient    proto.ElasticAgentStoreClient
+	artifactClient proto.ElasticAgentArtifactClient
+	logClient      proto.ElasticAgentLogClient
+
+	// overridden in tests to make fast
+	minCheckTimeout time.Duration
+}
+
+// NewV2 creates a client connection to Elastic Agent over the V2 control protocol.
+func NewV2(target string, token string, versionInfo VersionInfo, opts ...grpc.DialOption) V2 {
+	return &clientV2{
+		target:          target,
+		opts:            opts,
+		token:           token,
+		versionInfo:     versionInfo,
+		kickCh:          make(chan struct{}, 1),
+		errCh:           make(chan error),
+		unitsCh:         make(chan UnitChanged),
+		minCheckTimeout: CheckinMinimumTimeout,
+	}
+}
+
+// Start starts the connection to Elastic Agent.
+func (c *clientV2) Start(ctx context.Context) error {
+	c.ctx, c.cancel = context.WithCancel(ctx)
+	conn, err := grpc.DialContext(ctx, c.target, c.opts...)
+	if err != nil {
+		return err
+	}
+	c.client = proto.NewElasticAgentClient(conn)
+	c.storeClient = proto.NewElasticAgentStoreClient(conn)
+	c.artifactClient = proto.NewElasticAgentArtifactClient(conn)
+	c.logClient = proto.NewElasticAgentLogClient(conn)
+	c.startCheckin()
+	c.startActions()
+	return nil
+}
+
+// Stop stops the connection to Elastic Agent.
+func (c *clientV2) Stop() {
+	if c.cancel != nil {
+		c.cancel()
+		c.wg.Wait()
+		c.ctx = nil
+		c.cancel = nil
+	}
+}
+
+// UnitChanges returns channel client send unit change notifications to.
+func (c *clientV2) UnitChanges() <-chan UnitChanged {
+	return c.unitsCh
+}
+
+// Errors returns channel of errors that occurred during communication.
+func (c *clientV2) Errors() <-chan error {
```

Comment:
Can you document that failing to read `Errors()` or `UnitChanges()` will cause the client to block internally?

Reply:
It is documented on the interface.